File: workspace/test/code/load_greeting_test.dart
import 'package:mocktail/mocktail.dart';
import 'package:saying_hello/code/load_greeting.dart';
import 'package:saying_hello/contracts/greeting_api.dart';
import 'package:saying_hello/contracts/greeting_response.dart';
import 'package:test/test.dart';

class MockGreetingApi extends Mock implements GreetingApi {}

void main() {
  test('returns the personalized greeting for a non-empty name', () async {
    final api = MockGreetingApi();
    when(
      () => api.getGreeting('Ada'),
    ).thenAnswer((_) async => const GreetingResponse(message: 'Hello, Ada!'));

    final result = await loadGreeting('Ada', api);

    expect(result.submittedName, 'Ada');
    expect(result.message, 'Hello, Ada!');
  });

  test('trims the name before calling the API', () async {
    final api = MockGreetingApi();
    when(
      () => api.getGreeting('Ada'),
    ).thenAnswer((_) async => const GreetingResponse(message: 'Hello, Ada!'));

    final result = await loadGreeting('  Ada  ', api);

    expect(result.submittedName, 'Ada');
    expect(result.message, 'Hello, Ada!');
  });

  test('returns a friendly message when the API is unavailable', () async {
    final api = MockGreetingApi();
    when(() => api.getGreeting('Ada')).thenThrow(Exception('network error'));

    final result = await loadGreeting('Ada', api);

    expect(result.submittedName, 'Ada');
    expect(result.message, 'Sorry, the greeting API is unavailable right now.');
  });
}

File: workspace/src/code/load-greeting.test.ts
import { describe, expect, it, vi } from "vitest";

import type { GreetingApi } from "../contracts/greeting-api";
import { loadGreeting } from "./load-greeting";

describe("loadGreeting", () => {
  it("returns the personalized greeting for a non-empty name", async () => {
    const getGreeting = vi.fn().mockResolvedValue({ message: "Hello, Ada!" });
    const api: GreetingApi = { getGreeting };

    const result = await loadGreeting("Ada", api);

    expect(getGreeting).toHaveBeenCalledWith("Ada");
    expect(result).toEqual({
      submittedName: "Ada",
      message: "Hello, Ada!",
    });
  });

  it("trims the name before calling the API", async () => {
    const getGreeting = vi.fn().mockResolvedValue({ message: "Hello, Ada!" });
    const api: GreetingApi = { getGreeting };

    const result = await loadGreeting("  Ada  ", api);

    expect(getGreeting).toHaveBeenCalledWith("Ada");
    expect(result).toEqual({
      submittedName: "Ada",
      message: "Hello, Ada!",
    });
  });

  it("returns a friendly message when the API is unavailable", async () => {
    const getGreeting = vi.fn().mockRejectedValue(new Error("network error"));
    const api: GreetingApi = { getGreeting };

    const result = await loadGreeting("Ada", api);

    expect(getGreeting).toHaveBeenCalledWith("Ada");
    expect(result).toEqual({
      submittedName: "Ada",
      message: "Sorry, the greeting API is unavailable right now.",
    });
  });
});
